# Part 1. Equation of a Slime

How many late days are you using for this assignment?
0


In [16]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
import numpy as np

## 1. Loading the dataset

In [17]:
# Using pandas load the dataset
sciDat = pd.read_csv('science_data_large.csv')
# Output the first 15 rows of the data
pd.DataFrame(sciDat.head(15))



,Temperature °C,Mols KCL,Size nm^3
0,469,647,6.244743e+05
1,403,694,5.779610e+05
2,302,975,6.196847e+05
3,779,916,1.460449e+06
4,901,18,4.325726e+04
5,545,637,7.124634e+05
6,660,519,7.006960e+05
7,143,869,2.718260e+05
8,89,461,8.919803e+04
9,294,776,4.770210e+05


In [18]:
# Display a summary of the table information (number of datapoints, etc.)
sciDat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


## 2. Splitting the dataset

In [19]:
# Take the pandas dataset and split it into our features (X) and label (y)
feat = sciDat[["Temperature °C","Mols KCL"]]
lab = sciDat["Size nm^3"]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
xTr, xTe, yTr, yTe = train_test_split(feat, lab, test_size=.1, train_size=.9, random_state=42)
xTr.shape, xTe.shape, yTr.shape, yTe.shape


((900, 2), (100, 2), (900,), (100,))

## 3. Perform a Linear Regression

In [20]:
# Use sklearn to train a model on the training set
reg = LinearRegression().fit(xTr, yTr)

# Create a sample datapoint and predict the output of that sample with the trained model
sInd = np.random.randint(0, len(xTe))
print("Sample Index: " + str(sInd))
sample = xTe.iloc[[sInd]]
print("Sample: " + str(sample))
sampleO = reg.predict(sample)
print("Prediction: " + str(sampleO[0]))
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
print("Score: " + str(reg.score(xTe, yTe)))

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
print("Coefficient 1: " + str(reg.coef_[0]))
print("Coefficient 2: " + str(reg.coef_[1]))
print("Intercept: " + str(reg.intercept_))

Sample Index: 77
Sample:     Temperature °C  Mols KCL
39             297       126
Prediction: -22026.416434736806
Score: 0.8552472077276095
Coefficient 1: 866.1464133719209
Coefficient 2: 1032.6950664857964
Intercept: -409391.47958340764


Write the linear equation of a slime: (example equation: $E = mc^2$)

$S = 866.14641x_1 + 1032.69507x_2- 409391.47958$

Where $x_1$ is the Temp and $x_2$ is Mols

Report on score and explain meaning:
The score is essentially a score telling how well the lin. reg. equation and the data actually fit eachother. In more specific terms, it is the proportion of variance between the independent and dependent variables. 

## 4. Use Cross Validation

In [21]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
kf = KFold(n_splits=5, random_state=42, shuffle=True)
cvs = cross_val_score(reg, feat, lab, cv = kf)

# Report on their finding and their significance
print("CVS Mean: " + str(cvs.mean()))
print("CVS Standard Deviation: " + str(cvs.std()))
print("CVS Scores: " + str(cvs))


CVS Mean: 0.8597294202684646
CVS Standard Deviation: 0.01838773713930639
CVS Scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]


Write findings here:

Our cross validation scores for 5 folds are roughly 0.86152, 0.82742, 0.87195, 0.88166, and 0.85609 which have a mean of about .85973 which is quite close to our test score of about 0.85525. But, our standard deviation makes up for this difference, and thus, our original test falls well within the bounds of what was expected. Therefore we can conclude our model is not over/under fit and makes relatively reliable predictions.

## 5. Using Polynomial Regression

In [22]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(2)
xPoly = poly.fit_transform(feat)
linReg = LinearRegression()


# Perform k-fold cross validation (as above)
cvs = cross_val_score(linReg, xPoly, lab, cv = kf)

# Report on the metrics and output the resultant equation as you did in Part 3.
print("CVS Mean: " + str(cvs.mean()))
print("CVS Standard Deviation: " + str(cvs.std()))
print("CVS Scores: " + str(cvs))


linReg.fit(xPoly, lab)


print("Coefficients: " + str(linReg.coef_))
print("Intercept: " + str(linReg.intercept_))
print("Score: " + str(linReg.score(xPoly, lab)))

CVS Mean: 1.0
CVS Standard Deviation: 0.0
CVS Scores: [1. 1. 1. 1. 1.]
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.23110940e-07 -1.05650504e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 1.657102257013321e-05
Score: 1.0


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$S = 12x_1-(1.23111*10^{-7})x_2+(1.05651*10^{-11})x_1^2+2(x_1* x_2)+(2.85714*10^{-2})x_2^2$

Report on the score and interpret: The score was 1 for every test run. This means that the equation generated for the given data set is perfect and will always predict the correct size given Temp and Mols. 

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [23]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
import sklearn.neural_network as nn

In [24]:
# Load the dataset. Then train and evaluate the classification models.
# Using pandas load the dataset
ckdDat = pd.read_csv('ckd_feature_subset.csv')
feat = ckdDat[["age","bp","wbcc","appet_poor","appet_good","rbcc"]]
lab = ckdDat["Target_ckd"]

models = {
"Logistic Regression" : LogisticRegression(random_state=42),
"Support Vector Machines" : SVC(random_state=42),
"k-Nearest Neighbors" : KNeighborsClassifier(),
"Neural Networks" : nn.MLPClassifier(max_iter=1000, random_state=42)
}

results = []

for modelN, model in models.items():
    cvs = cross_val_score(model, feat, lab, cv = kf)
    results.append([modelN,str(cvs.mean()), str(cvs.std())])

pd.DataFrame(results, columns = ["Model", "Mean Accuracy", "Standard Deviation"])
    

,Model,Mean Accuracy,Standard Deviation
0,Logistic Regression,0.8565591397849461,0.0662686697755876
1,Support Vector Machines,0.9281720430107526,0.04760126519964333
2,k-Nearest Neighbors,0.9279569892473118,0.05243961560473918
3,Neural Networks,0.9350537634408603,0.040465552982107675


## Results and Conclusion for Classification Experiments

The performance of four different machine learning models was evaluated, with results indicating that Neural Networks achieved the highest mean accuracy of approximately 93.51%, followed closely by Support Vector Machines at 92.82% and k-Nearest Neighbors at 92.80%. Logistic Regression provided a solid baseline with an accuracy of around 85.66%, though it lagged behind the others. Notably, Neural Networks exhibited the lowest standard deviation (0.0405), suggesting the most consistent performance across cross-validation folds. Overall, this analysis highlights the superiority of complex models like Neural Networks in classification tasks while emphasizing the importance of stability in model performance.

In [25]:
# Experiments with Neural Network.

def evaluate_model(model): 
    cvs = cross_val_score(model, feat, lab, cv=kf)
    return cvs.mean(), cvs.std()

layer_variations = {
    'Single Layer 100': nn.MLPClassifier(hidden_layer_sizes=(100,), max_iter=2000, random_state=42),
    'Two Layers 100, 200': nn.MLPClassifier(hidden_layer_sizes=(100, 200), max_iter=2000, random_state=42),
}

solver_variations = {
    'Adam': nn.MLPClassifier(solver='adam', max_iter=2000, random_state=42),
    'SGD': nn.MLPClassifier(solver='sgd', max_iter=2000, random_state=42),
    'LBFGS': nn.MLPClassifier(solver='lbfgs', max_iter=2000, random_state=42),
}

activation_variations = {
    'ReLU': nn.MLPClassifier(activation='relu', max_iter=2000, random_state=42),
    'Tanh': nn.MLPClassifier(activation='tanh', max_iter=2000, random_state=42),
    'Logistic': nn.MLPClassifier(activation='logistic', max_iter=2000, random_state=42),
}

results = {}
for variation in [layer_variations, solver_variations, activation_variations]:
    for name, model in variation.items():
        results[name] = evaluate_model(model)

pd.DataFrame(results, index=['Mean Accuracy', 'Standard Deviation']).T


,Mean Accuracy,Standard Deviation
Single Layer 100,0.935054,0.040466
"Two Layers 100, 200",0.967527,0.035340
Adam,0.935054,0.040466
SGD,0.849892,0.059545
LBFGS,0.980430,0.015984
ReLU,0.935054,0.040466
Tanh,0.941505,0.037327
Logistic,0.915269,0.048238


## Results and Conclusion for Neural Network Experiments

I did a pretty exhaustive test of the possible variations you can create using the SKLearn neural net. In my findings, i discovered that when you increase the complexity, or neuron count, as one may expect the accuracy goes up. When it comes to solvers, I found that the LBFGS optimization algorithm provided the most accurate outputs. This may be because it is made specifically for smaller datasets (which this one very much is) so that definitely makes sense. Finally, with activation functions Tanh performed the best of the 3 provided. 

Below is a test with all 3 of the best options just because im curious.  

In [26]:
nnLol= nn.MLPClassifier(hidden_layer_sizes=(100, 200), solver='lbfgs', activation= "tanh", max_iter=2000, random_state=42)
pd.DataFrame(evaluate_model(nnLol), ['Mean Accuracy', 'Standard Deviation'])


,0
Mean Accuracy,0.973978
Standard Deviation,0.037658
